File: scripts/algorithms/L/Longest Absolute File Path/Longest Absolute File Path.cpp
// Runtime: 0 ms (Top 100.00%) | Memory: 6.5 MB (Top 50.23%)
// Using Map O(300 + N)

class Solution {
    public:
    int lengthLongestPath(string input) {
        input.push_back('\n');
        vector<int> levels(301, 0);
        int ans = 0;
        int curr_tabs = 0;
        bool is_file = false;
        int curr_word_len = 0;
        int total_len = 0;

        for(char c : input)
        {
            if(c == '\t')
            {
                curr_tabs++;
            }
            else if(c == '\n')
            {
                if(curr_tabs == 0)
                {
                    levels[0] = curr_word_len;
                }
                else
                {
                    levels[curr_tabs] = levels[curr_tabs-1] + curr_word_len;
                }

                if(is_file)
                {
                    ans = max(ans, levels[curr_tabs] + curr_tabs);
                    // levels[curr_tabs] = 0;
                }
                curr_tabs = 0;
                is_file = false;
                curr_word_len = 0;
            }
            else if(c == '.')
            {
                is_file = true;
                curr_word_len++;
            }
            else
            {
                curr_word_len++;
            }
        }
        return ans;
    }
};

File: scripts/algorithms/L/Longest Absolute File Path/Longest Absolute File Path.rs
// Runtime: 0 ms (Top 100.0%) | Memory: 2.20 MB (Top 83.33%)

impl Solution {
    pub fn length_longest_path(input: String) -> i32 {
        let split: Vec<&str> = input.split('\n').collect();

        let mut stack: Vec<(usize, &str)> = vec![];
        let mut ans = 0;

        for s in split {
            let mut lvl = 0;
            for ch in s.chars() {
                if ch == '\t' {
                    lvl += 1;
                } else {
                    break;
                }
            }
            let fname = &s[(lvl as usize)..];

            while stack.len() > 0 && stack[stack.len() - 1].0 >= lvl {
                stack.pop();
            }
            stack.push((lvl, fname.clone()));

            if fname.contains('.') {
                let v = stack
                    .clone()
                    .into_iter()
                    .map(|(_, s)| s.clone().to_string())
                    .collect::<Vec<String>>()
                    .join("/");
                    
                ans = ans.max(v.len() as i32);
            }
        }

        ans
    }
}
